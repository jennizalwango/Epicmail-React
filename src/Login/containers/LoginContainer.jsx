import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Login from '../components/LoginComponent';
import loginUtil from '../redux/validations';
import loginAction from '../redux/actions/LoginAction';

export class LoginContainer extends Component {
  constructor(props) {
    super(props);
    this.state = {
      email: '',
      password: '',
      errors: {
        invalid_email: false,
      },
    };
  }

  onChange = (event) => {
    this.setState({ [event.target.name]: event.target.value });

    const { target } = event;
    const { name, value } = target;
    const { errors } = this.state;
    switch (name) {
      case 'email':
        if (!this.checkValidEmail(value)) {
          this.setState({ errors: { userEmail: 'Please provide a valid email address' } });
          errors.invalid_email = true;
        } else {
          this.setState({ errors: { userEmail: '' } });
          errors.invalid_email = false;
        }
        break;
      default:
        break;
    }

    this.setState({ [event.target.name]: event.target.value });
  }

  checkValidEmail = (userEmail) => {
    const mailFormat = loginUtil.LOGIN_UTILS;
    if (userEmail.match(mailFormat)) {
      return true;
    }
    return false;
  };

  handleSubmit = (event) => {
    event.preventDefault();
    const { email, password } = this.state;
    const userdata = {
      user: {
        email,
        password,
      },
    };

    if (email === '') {
      this.setState({ errors: { userEmail: 'Email field required' } });
    }
    if (password === '') {
      this.setState({ errors: { userPassword: 'Password field is required' } });
    }
    const { propsLoginAction } = this.props;

    const { errors } = this.state;
    if (!errors.invalid_email && errors.userEmail === '') {
      propsLoginAction(userdata, this.props);
    }
  }

  render() {
    const { email, password, errors } = this.state;
    const {
      isLoggingIn,
      isLoading,
      loginFailed,
    } = this.props;
    return (
      <Login
        handleSubmit={this.handleSubmit}
        onChange={this.onChange}
        email={email}
        password={password}
        errors={errors}
        isLoggingIn={isLoggingIn}
        loginFailed={loginFailed}
        isLoading={isLoading}
      />
    );
  }
}

LoginContainer.propTypes = {
  propsLoginAction: PropTypes.func.isRequired,
  isLoading: PropTypes.bool.isRequired,
  isLoggingIn: PropTypes.bool.isRequired,
  loginFailed: PropTypes.bool.isRequired,
};

const mapStateToProps = state => ({
  login: state.login,
  isLoggingIn: state.login.is_logging_in,
  loginFailed: state.login.login_failed,
  isLoading: state.login.isLoading,
});
export default connect(mapStateToProps, { propsLoginAction: loginAction })(LoginContainer);
