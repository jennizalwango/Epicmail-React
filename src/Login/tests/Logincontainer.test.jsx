import React from 'react';
import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';
import { LoginContainer } from '../containers/LoginContainer';

describe('loginContainer Component', () => {
  const props = {
    onChange: jest.fn(),
    handleSubmit: jest.fn(),
    checkValidEmail: jest.fn(),
    propsLoginAction: jest.fn(),
    isLoading: false,
    isLoggingIn: false,
    loginFailed: false,
  };

  const event = {
    preventDefault: () => {},
    target: {
      name: 'email',
      value: 'abc',
    },
  };
  const wrapper = shallow(<LoginContainer {...props} />);

  it('matches snapshot', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should call an onChange function', () => {
    const component = shallow(
      <LoginContainer {...props} />,
    );
    component.setState({ email: 'email' });
    component.instance().onChange(event);
    expect(component.instance().state.email).toBe('abc');
  });

  it('should call handleSubmit', () => {
    const component = shallow(
      <LoginContainer {...props} />,
    );
    component.setState({ email: 'jane@us.com', password: '' });
    component.instance().handleSubmit(event);
    expect(component.state('errors').userPassword).toBe('Password field is required');
  });

  it('should call a submit a form', () => {
    wrapper.setState({
      email: 'jennyj',
    });
    const userEmail = {
      match: jest.fn(),
    };
    const instance = wrapper.instance();
    const checkValidEmail = jest.spyOn(instance, 'checkValidEmail');
    checkValidEmail(userEmail);
    instance.handleSubmit({ preventDefault: jest.fn() });
    expect(checkValidEmail).toHaveBeenCalled();
  });

  it('Valid Email', () => {
    wrapper.setState({
      email: 'jenny@gmail.com',
    });
    const userEmail = {
      match: jest.fn(),
    };
    const instance = wrapper.instance();
    const checkValidEmail = jest.spyOn(instance, 'checkValidEmail');
    instance.handleSubmit({ preventDefault: jest.fn() });
    checkValidEmail(userEmail);
    expect(checkValidEmail).toHaveBeenCalled();
  });

  it('should call submit form', () => {
    wrapper.setState({
      email: 'jennyj',
    });
    const userEmail = {
      match: jest.fn(),
    };
    const instance = wrapper.instance();
    const checkValidEmail = jest.spyOn(instance, 'checkValidEmail');
    instance.handleSubmit({ preventDefault: jest.fn() });
    checkValidEmail(userEmail);
    expect(checkValidEmail).toHaveBeenCalled();
  });

  it('should validate an empty password', () => {
    const instance = wrapper.instance();
    instance.handleSubmit({ preventDefault: jest.fn() });
    expect('Password field is required').toEqual(instance.state.errors.userPassword);
  });

  it('should call loginAction when handling submit', () => {
    wrapper.setState({
      email: 'jennyzalwango12@gmail.com',
      password: 'password',
      errors: {},
    });
    const instance = wrapper.instance();
    instance.handleSubmit({ preventDefault: jest.fn() });
    props.propsLoginAction();
    expect(props.propsLoginAction).toHaveBeenCalled();
  });

  it('should load the component', () => {
    expect(wrapper).toHaveLength(1);
  });
});
