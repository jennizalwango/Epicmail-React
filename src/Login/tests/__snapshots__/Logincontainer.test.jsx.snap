// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loginContainer Component matches snapshot 1`] = `
<Login
  email=""
  errors={
    Object {
      "invalid_email": false,
    }
  }
  handleSubmit={[Function]}
  isLoading={false}
  isLoggingIn={false}
  loginFailed={false}
  onChange={[Function]}
  password=""
/>
`;
